content/blog/2019-11-18-edge-architecture.md

```diff
@@ -0,0 +1,72 @@
+
+To calculate the final penalty, the algorithm sums up both intra- and intro-contention for ALUs. In addition, it introduces two factors, which are fullness and criticality, respectively. For the former, it corresponds to how full a block it. For the latter, it compares the maximum path length of an instruction with the length of the critical path. The final placement cost is just a combination of all the above numbers.
+
+#### Global Register Prioritization
+
+The main idea of this heuristic is performing loop-related optimizations. The algorithm performs the following three heuristics. First, smaller loops (in terms of the number of blocks) will be scheduled first. Second, loop-carried dependencies will be scheduled first. Third, when calculating the longest path, the algorithm will also consider the predecessor and the successor of the current block.
+
+#### Path Volume Scheduling
+
+The goal here is to find the best path from a source unit to a destination unit. This is not as simple as it seems because some units might be fully occupied by other instructions already. In addition, we need to fit all intermediate instructions to the path we find. The authors apply a depth-first search with **iterative deepening**. The idea is to set an upper bound first. It will keep increasing the upper bound until it finds a feasible solution.
+
+#### Putting It All Together
+
+Finally, we can combine all heuristics into a single function that calculates the placement cost. Namely, we first consider the penalty brought by the contention. Then, we consider the loop-related optimizations. Finally, we include the routing cost derived by the pathfinding algorithm.
+
+## Evaluation
+
+The authors selected a wide range of benchmarks with different levels of concurrency and memory behavior. Since dense blocks require more algorithmic work, the authors modified the selected benchmarks by hand. Then they compared the results with the baseline GRST algorithm and the simulated annealing results. They also show the step by step results after applying different heuristics. In conclusion, their SPS algorithm improves the cycle count by over 21% in average comparing with GRST. Moreover, the SPS results are within 5% difference in average comparing with the simulated annealing results. In addition, the authors performed a cross validation by performing the same algorithm on the unmodified benchmarks. The results still demonstrate an over 17% improvement in average comparing with GRST.
```

Comment:
It would be useful to break down a little bit more about those "step by step" results, showing which parts of the heuristics mattered more.